qiskit/opflow/primitive_ops/tapered_pauli_sum_op.py

```diff
@@ -0,0 +1,64 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+""" TaperedPauliSumOp Class """
+
+from typing import Union
+
+from qiskit.circuit import ParameterExpression
+from qiskit.quantum_info import SparsePauliOp
+
+from ..converters.z2_symmetries import Z2Symmetries
+from ..operator_base import OperatorBase
+from .pauli_sum_op import PauliSumOp
+
+
+class TaperedPauliSumOp(PauliSumOp):
+    """Class for PauliSumOp after tapering"""
+
+    def __init__(
+            self,
+            primitive: SparsePauliOp,
+            z2_symmetries: Z2Symmetries,
+            coeff: Union[int, float, complex, ParameterExpression] = 1.0,
+    ) -> None:
+        """
+        Args:
+            primitive: The SparsePauliOp which defines the behavior of the underlying function.
+            z2_symmetries: Z2 symmetries which the Operator has.
+            coeff: A coefficient multiplying the primitive.
+
+        Raises:
+            TypeError: invalid parameters.
+        """
+        super().__init__(primitive, coeff)
+        self._z2_symmetries = z2_symmetries
+
+    @property
+    def z2_symmetries(self) -> Z2Symmetries:
+        """
+        Z2 symmetries which the Operator has.
+
+        Returns:
+            The Z2 Symmetries.
+        """
+        return self._z2_symmetries
+
+    def equals(self, other: OperatorBase):
```

Comment:
I am wondering about this equals. Having the z2_symmetries in the operator was convenient before (in the old legacy logic). But it was a simply a field in the 'normal' operator (the WeightedPauliOperator) that was accessible when you needed it not a whole new sub-type. If there were no Z2 symmetries it returned an value to that effect. Here in this sub-type, with equality being overridden and including the z2_symmetries I am wondering about normal usage. Even though this operator was generated via a symmetry operation, and maybe you want to access that information, somehow if you what you declare you support is a PauliSumOp class and get passed some instances (those and this subtype) if you want to check if the operators are equal, which normally means identical Pauli sums, it will fail (return False), even in cases when they are all the same if the z2 symmetry they were derived from is not.  Now to get the behavior you want you would have to cast back to PauliSumOp which may not be so obvious to most people.

Reply:
TaperedPauliSumOp is a class that preserves the history of tapering. For example, even if the matrices are the same as operators, they are different operators if they belong to different sectors (tapering_value). If the equality comparison should not include Z2Symmetries, then the design of operator to hold Z2Symmetries as an instance variable (property) is wrong in the first place. In that case, I would revert and adopt my first design. If you want to do the comparison you want, you can convert it to PauliSumOp.

Reply:
I think I would have had a field in the PauliSumOp for this, akin to what the grouping. And for grouping, its an extra field that is not counted in equality in PauliSumOp. Do we really care about the 'history' when doing equality - my take would have been more that the equality in that regard would be unusual. My main concern here is that chemistry can create a bunch of tapered operators and its convenient that this is carried with them in case you want to use it. Most often when we consumed the operator we did not care over its history of creation. With how you have it one can just always cast to PauliSumOp for the equals - hopefully people remember to do this when consuming the operator for equality. With this as a derived class I guess there are other operations that are not going to respect this field, like multiply by some constant for example. Maybe you are right and it should be kept out of band by anyone who needs it - all I know is that we had this in the past the and then any user needed to maintain these two separate things. Hence in one of the refactors we added it as a field in the operator, which was set to None by default - we never had a separate derived class.

Reply:
OK, I decided to remove `__eq__` from TaperedPauliSumOp.